js/navigation.js

```diff
@@ -5,108 +5,95 @@
-		menu.className += ' nav-menu';
+	if ( ! menu.classList.contains( 'nav-menu' ) ) {
+		menu.classList.add( 'nav-menu' );
 	}
 
-	button.onclick = function() {
-		if ( -1 !== container.className.indexOf( 'toggled' ) ) {
-			container.className = container.className.replace( ' toggled', '' );
+	// Toggle the .toggled class and the aria-expanded value each time the button is clicked.
+	button.addEventListener( 'click', function() {
+		siteNavigation.classList.toggle( 'toggled' );
+
+		if ( button.getAttribute( 'aria-expanded' ) === 'true' ) {
 			button.setAttribute( 'aria-expanded', 'false' );
 		} else {
-			container.className += ' toggled';
 			button.setAttribute( 'aria-expanded', 'true' );
 		}
-	};
+	} );
 
-	// Close small menu when user clicks outside
+	// Remove the .toggled class and set aria-expanded to false when the user clicks outside the navigation.
 	document.addEventListener( 'click', function( event ) {
-		var isClickInside = container.contains( event.target );
+		const isClickInside = siteNavigation.contains( event.target );
 
 		if ( ! isClickInside ) {
-			container.className = container.className.replace( ' toggled', '' );
+			siteNavigation.classList.remove( 'toggled' );
 			button.setAttribute( 'aria-expanded', 'false' );
 		}
 	} );
 
 	// Get all the link elements within the menu.
-	links = menu.getElementsByTagName( 'a' );
+	const links = menu.getElementsByTagName( 'a' );
+
+	// Get all the link elements with children within the menu.
+	const linksWithChildren = menu.querySelectorAll( '.menu-item-has-children > a, .page_item_has_children > a' );
+
+	// Toggle focus each time a menu link is focused or blurred.
+	for ( const link of links ) {
+		link.addEventListener( 'focus', toggleFocus, true );
+		link.addEventListener( 'blur', toggleFocus, true );
+	}
 
-	// Each time a menu link is focused or blurred, toggle focus.
-	for ( i = 0, len = links.length; i < len; i++ ) {
-		links[i].addEventListener( 'focus', toggleFocus, true );
-		links[i].addEventListener( 'blur', toggleFocus, true );
+	// Toggle focus each time a menu link with children receive a touch event.
+	for ( const link of linksWithChildren ) {
+		link.addEventListener( 'touchstart', toggleFocus, false );
 	}
 
 	/**
 	 * Sets or removes .focus class on an element.
 	 */
 	function toggleFocus() {
-		var self = this;
-
-		// Move up through the ancestors of the current link until we hit .nav-menu.
-		while ( -1 === self.className.indexOf( 'nav-menu' ) ) {
-			// On li elements toggle the class .focus.
-			if ( 'li' === self.tagName.toLowerCase() ) {
-				if ( -1 !== self.className.indexOf( 'focus' ) ) {
-					self.className = self.className.replace( ' focus', '' );
-				} else {
-					self.className += ' focus';
+		if ( event.type === 'focus' || event.type === 'blur' ) {
+			let self = this;
+			// Move up through the ancestors of the current link until we hit .nav-menu.
+			while ( ! self.classList.contains( 'nav-menu' ) ) {
```

Comment:
This while loop looks odd. Could we just say something like `self = this.closest( '.nav-menu' )`?  Also, toggling a class name on all the li elements between the current item and the nav menu seems surprising. Is the goal to remove the class name or add it?

Reply:
It's a bit confusing but it's the copy of the file from `_s`; we took this file from there originally, and this is the most recent released version of the navigation code. Since the focus navigation functionality does still appear to work, I was inclined to treat it as an upstream dependency and didn't investigate some of the confusing bits too deeply.